openmc/material.py

```diff
@@ -0,0 +1,1109 @@
+    using the :attr:`Cell.fill` attribute.
+
+    Parameters
+    ----------
+    material_id : int, optional
+        Unique identifier for the material. If not specified, an identifier will
+        automatically be assigned.
+    name : str, optional
+        Name of the material. If not specified, the name will be the empty
+        string.
+    temperature : float, optional
+        Temperature of the material in Kelvin. If not specified, the material
+        inherits the default temperature applied to the model.
+
+    Attributes
+    ----------
+    id : int
+        Unique identifier for the material
+    temperature : float
+        Temperature of the material in Kelvin.
+    density : float
+        Density of the material (units defined separately)
+    density_units : str
+        Units used for `density`. Can be one of 'g/cm3', 'g/cc', 'kg/m3',
+        'atom/b-cm', 'atom/cm3', 'sum', or 'macro'.  The 'macro' unit only
+        applies in the case of a multi-group calculation.
+    depletable : bool
+        Indicate whether the material is depletable.
+    nuclides : list of tuple
+        List in which each item is a 3-tuple consisting of a nuclide string, the
+        percent density, and the percent type ('ao' or 'wo').
+    isotropic : list of str
+        Nuclides for which elastic scattering should be treated as though it
+        were isotropic in the laboratory system.
+    average_molar_mass : float
+        The average molar mass of nuclides in the material in units of grams per
+        mol.  For example, UO2 with 3 nuclides will have an average molar mass
+        of 270 / 3 = 90 g / mol.
+    volume : float
+        Volume of the material in cm^3. This can either be set manually or
+        calculated in a stochastic volume calculation and added via the
+        :meth:`Material.add_volume_information` method.
+    paths : list of str
+        The paths traversed through the CSG tree to reach each material
+        instance. This property is initialized by calling the
+        :meth:`Geometry.determine_paths` method.
+    num_instances : int
+        The number of instances of this material throughout the geometry.
+    fissionable_mass : float
+        Mass of fissionable nuclides in the material in [g]. Requires that the
+        :attr:`volume` attribute is set.
+
+    """
+
+    next_id = 1
+    used_ids = set()
+
+    def __init__(self, material_id=None, name='', temperature=None):
+        # Initialize class attributes
+        self.id = material_id
+        self.name = name
+        self.temperature = temperature
+        self._density = None
+        self._density_units = 'sum'
+        self._depletable = False
+        self._paths = None
+        self._num_instances = None
+        self._volume = None
+        self._atoms = {}
+        self._isotropic = []
+
+        # A list of tuples (nuclide, percent, percent type)
+        self._nuclides = []
```

Comment:
Here and elsewhere, consider making tuples into [`collections.namedtuple`s](https://docs.python.org/3.8/library/collections.html#collections.namedtuple) rather than relying on the indexes. (I've only recently started converting my own projects to this and I've uncovered more than 1 latent bug caused by forgetting the ordering of tuple elements)